lib/logflare/backends/adaptor/datadog_adaptor.ex

```diff
@@ -0,0 +1,190 @@
+defmodule Logflare.Backends.Adaptor.DatadogAdaptor do
+  @moduledoc """
+  Wrapper module for `Logflare.Backends.Adaptor.WebhookAdaptor` to provide API
+  for DataDog logs ingestion endpoint.
+  """
+
+  defmodule Client do
+    use GenServer
+
+    require Logger
+
+    # - `queue` - list of elements in queue
+    # - `size` - amount of elements in queue (cached to not use `length/1` all
+    #   the time)
+    # - `threshold` - maximal amount of items in queue before sending
+    # - `timeout` - time to wait before sending all messages in queue
+    # - `timer` - reference to timer used
+    defstruct queue: [],
+              size: 0,
+              threshold: 100,
+              timeout: 1000,
+              timer: nil,
+              url: nil,
+              method: :post,
+              headers: []
+
+    def start_link(opts), do: GenServer.start_link(__MODULE__, Map.new(opts))
+
+    def ingest(_pid, []), do: :ok
+    def ingest(pid, logs), do: GenServer.cast(pid, {:events, logs})
+
+    @impl GenServer
+    def init(opts) do
+      {:ok,
+       %__MODULE__{
+         threshold: opts[:threshold] || 100,
+         timeout: opts[:timeout] || 1000,
+         method: opts[:method] || :post,
+         url: Map.fetch!(opts, :url),
+         headers: Enum.to_list(opts[:headers] || [])
+       }}
+    end
+
+    @impl GenServer
+    def handle_cast({:events, events}, %__MODULE__{} = state) do
+      len = length(events)
+
+      new_state = %__MODULE__{
+        state
+        | queue: events ++ state.queue,
+          size: len + state.size,
+          timer: start_timer(state.timer, state.timeout)
+      }
+
+      if new_state.size >= new_state.threshold do
+        {:noreply, new_state, {:continue, :send}}
+      else
+        {:noreply, new_state}
+      end
+    end
+
+    @impl GenServer
+    def handle_info(:timeout, state) do
+      {:noreply, state, {:continue, :send}}
+    end
+
+    @impl GenServer
+    def handle_continue(:send, state) do
+      send_logs(state)
+      {:noreply, %__MODULE__{state | queue: [], size: 0, timer: cancel_timer(state.timer)}}
+    end
+
+    defp start_timer(nil, timeout) do
+      {:ok, ref} = :timer.send_after(timeout, :timeout)
+
+      ref
+    end
+
+    defp start_timer(ref, _timeout), do: ref
+
+    defp cancel_timer(nil), do: nil
+
+    defp cancel_timer(ref) do
+      :timer.cancel(ref)
+      nil
+    end
+
+    defp send_logs(%__MODULE__{} = state) do
+      middlewares = [
+        Tesla.Middleware.Telemetry,
+        Tesla.Middleware.JSON
+      ]
+
+      client = Tesla.client(middlewares)
+
+      Task.start(fn ->
+        case Tesla.request(client,
```

Comment:
retry with backoff would be great

Reply:
we can add it after we add in batching for webhook adaptor

Reply:
Yeah. That module probably should be extracted to be independent of the DataDog adaptor, or even to separate library, but for now I wanted to have everything simple and in single place.